File: typescript/tests/booking/application/integration/payment/adapter/refundRequest.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { bookingRefused } from '@/booking/domain/events/bookingResultEvents.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import { refundRequestFromBookingRefused } from '@/booking/application/integration/payment/adapter/refundRequest.js';

describe('RefundRequest (booking -> payment)', () => {
  it('adapts BookingRefused to RefundRequestIntegrationCommand', () => {
    const bookingId = generateId((value) => new BookingId(value));
    const giftCardReference = crypto.randomUUID();
    const amount = new Money(30);
    const event = bookingRefused(bookingId, giftCardReference, amount);

    const command = refundRequestFromBookingRefused(event);

    expect(command.clientReference).toBe(bookingId.value.value);
    expect(command.amount).toBe(amount);
  });

  it('throws on null event', () => {
    expect(() => refundRequestFromBookingRefused(null as unknown as ReturnType<typeof bookingRefused>)).toThrow();
  });
});

File: java/src/test/java/it/giannibombelli/wsc2026/booking/application/integration/payment/adapter/RefundRequestTest.java
package it.giannibombelli.wsc2026.booking.application.integration.payment.adapter;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.payment.integration.RefundRequestIntegrationCommand;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class RefundRequestTest {

    @Test
    void fromBookingRefused_returnsIntegrationCommand() {
        BookingId bookingId = EntityId.generate(BookingId::new);
        UUID giftCardReference = UUID.randomUUID();
        Money amount = new Money(new BigDecimal("30.00"));
        BookingResultEvents.BookingRefused event = new BookingResultEvents.BookingRefused(bookingId, giftCardReference, amount);

        RefundRequestIntegrationCommand command = RefundRequest.fromBookingRefused(event);

        assertThat(command.clientReference()).isEqualTo(bookingId.value().toString());
        assertThat(command.amount()).isEqualTo(amount);
    }

    @Test
    void fromBookingRefused_withNullEvent_throwsException() {
        assertThatThrownBy(() -> RefundRequest.fromBookingRefused(null))
            .isInstanceOf(IllegalArgumentException.class);
    }
}
